Importing playlists from Apple Music

rimma-kubanova opened the pull request:
# Problem
* No Integration of Apple Music for importing playlists beforehand

# Action

- Front
1. Added new modal **ImportAppleMusicPlaylistModal** for importing playlists from Apple Music. It's similar to Spotify Modal
2. Declared types for Apple Music playlists, songs , attributes, and etc. in types.d.ts
3. Added call for two API endpoints (1. For fetching all user playlists, 2. For fetching tracks from given playlist) in ApiService

- Back
1. New API endpoints (1. For fetching all user playlists, 2. For fetching tracks from given playlist) in playlist_api.py)
2. Added import_from_apple_music() that calls **troi** to create a playlist


- [ ] Open a separate PR for `import_tracks_from_spotify_playlist` functions to clean up

p.s This PR will not work until the[ PR in troi ](https://github.com/metabrainz/troi-recommendation-playground/pull/141)will not be merged

I have tested it using `./test.sh` and it gives 4 errors (2 error for troi, since changes are not yet merged, 2 error for panda, and Idk why because I havent touched it)

pep8speaks commented:
Hello @rimma-kubanova! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`listenbrainz/metadata_cache/apple/client.py`](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/metadata_cache/apple/client.py):

> [Line 57:40](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/metadata_cache/apple/client.py#L57): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`listenbrainz/troi/import_ms.py`](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/troi/import_ms.py):

> [Line 20:1](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/troi/import_ms.py#L20): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

* In the file [`listenbrainz/webserver/views/playlist_api.py`](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/webserver/views/playlist_api.py):

> [Line 880:9](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/webserver/views/playlist_api.py#L880): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 885:1](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/webserver/views/playlist_api.py#L885): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 965:5](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/webserver/views/playlist_api.py#L965): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 967:1](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/webserver/views/playlist_api.py#L967): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 970:1](https://github.com/metabrainz/listenbrainz-server/blob/f959b5751f562781f13ee0fa3d3bf8cf90551940/listenbrainz/webserver/views/playlist_api.py#L970): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

MonkeyDo commented:
I see [two failing tests in CI](https://github.com/metabrainz/listenbrainz-server/actions/runs/9563033039/job/26360701075?pr=2910#step:6:2193) that seem related to the PR, complaining about import (and also two failing dumps tests unrelated to your PR):

```
ImportError while importing test module '/code/listenbrainz/listenbrainz/tests/integration/test_api.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
[...]
    from troi.patches.playlist_from_ms import ImportPlaylistPatch
E   ModuleNotFoundError: No module named 'troi.patches.playlist_from_ms'


ImportError while importing test module '/code/listenbrainz/listenbrainz/tests/integration/test_playlist_api.py'.
Traceback:
[...]
    from troi.patches.playlist_from_ms import ImportPlaylistPatch
E   ModuleNotFoundError: No module named 'troi.patches.playlist_from_ms'
```

amCap1712 commented on listenbrainz/metadata_cache/apple/client.py:
```diff
@@ -41,3 +41,16 @@ def get(self, url, params=None):
             response.raise_for_status()
+
+    def get_user_data(self, url, music_user_token, params=None):
+        with self._get_requests_session() as http:
```
This likely won't work because if a music user token expires it cannot be refreshed without manual user intervention. If the developer token expires, the music user token expires too. So I think you should just try once and if there is a 403 error return that from the playlist import endpoint and show a relevant error message on the frontend asking users to reconnect their account.

MonkeyDo reviewed (commented):
I have a couple of general comments, but unfortunately I am currently unable to test the functionality, lacking a payed apple music account.
I've enlisted the help of @anshg1214 to test it, he might have more comments.

rimma-kubanova commented on listenbrainz/webserver/views/playlist_api.py:
```diff
@@ -923,7 +938,7 @@ def import_playlist_from_spotify(service):
 @ratelimit()
 @api_listenstore_needed
-def import_tracks_from_spotify_to_playlist(service, playlist_id):
+def import_tracks_from_spotify_playlist(service, playlist_id):
```
Thanks for the feedback! I think for now I will leave them as separate endpoints, and remove the service arg. I will open a separate PR to handle that this later, since for now I am not yet sure how to better scale it as more music services will be added.

MonkeyDo replied:
Great. My only advice would be to note that down on a list somewhere, because I know for sure I will forget myself :p

anshg1214 commented on frontend/js/src/user/playlists/components/ImportAppleMusicPlaylistModal.tsx:
```diff
@@ -0,0 +1,204 @@
+      id="ImportSpotifyPlaylistModal"
+      tabIndex={-1}
+      role="dialog"
+      aria-labelledby="ImportSpotifyPlaylistLabel"
```
```suggestion
      aria-labelledby="ImportAppleMusicPlaylistLabel"
```

anshg1214 commented on frontend/js/src/user/playlists/components/ImportAppleMusicPlaylistModal.tsx:
```diff
@@ -0,0 +1,204 @@
+            </button>
+            <h4
+              className="modal-title"
+              id="ImportSpotifyPlaylistLabel"
```
```suggestion
              id="ImportAppleMusicPlaylistLabel"
```

anshg1214 commented on frontend/js/src/user/playlists/Playlists.tsx:
```diff
@@ -197,6 +199,29 @@ export default class UserPlaylists extends React.Component<
+                    data-toggle="modal"
+                    data-target="#ImportSpotifyPlaylistModal"
+                  >
+                    <FontAwesomeIcon icon={faMusic} />
```
You should use the apple logo here for consistency 
```suggestion
                    <FontAwesomeIcon icon={faApple} />
```

anshg1214 reviewed (changes requested):
The PR looks great! Just a couple of small changes.

Also, please have a look at the PEP 8 warnings above.

rimma-kubanova commented:
Hi, @anshg1214!

Thank you so much for your feedback! I have addressed all the issues you pointed out, so everything should be fine now. Regarding the error in Troi, I've already committed [my PR](https://github.com/metabrainz/troi-recommendation-playground/pull/142) to make the description optional. Once it's merged, it should be resolved.

amCap1712 reviewed (approved):
LGTM, thanks!

anshg1214 reviewed (approved):
LGTM!

TODO: Need to update the `troi` package in requirements.txt before merging

MonkeyDo commented on frontend/js/src/user/playlists/components/ImportAppleMusicPlaylistModal.tsx:
```diff
@@ -0,0 +1,208 @@
+                <p>
+                  Loading playlist {playlistLoading}... It might take some time
+                </p>
+                <Loader isLoading={!!playlistLoading} />
```
Not important at all, and not worth changing really, but for reference you could skip this conditional since you already have `!!playlistLoading` for conditional renderin at line 184
